### PR TITLE
[Example] Add Top-n-sigma logit truncation custom logits processor

### LINKED ISSUE #45023: [RFC] Add Top-nσ logit truncation example via custom logits processor
## RFC: Add Top-nσ Logit Truncation Example via Custom Logits Processor

**Reference**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### Motivation

Top-nσ is a logit-space dynamic truncation method proposed by Tang et al. (ACL 2025). Unlike probability-space filters such as `top_p` or `min_p`, it applies **before softmax** and keeps the candidate set less sensitive to temperature changes.

This issue proposes adding a **research-oriented example implementation** using vLLM's existing custom logits processor API. It is **not** a request to add a new built-in sampling parameter.

### Scope

This proposal only adds an **example file** demonstrating how to implement top-nσ as a custom logits processor. It:
- Does **not** add a new `SamplingParams` field
- Does **not** change any existing sampling behavior
- Does **not** modify the core `Sampler` class

### Algorithm

For each request row:

```
1. max_logit = logits.max()
2. std_logit = logits.std()
3. threshold = max_logit - n * std_logit
4. logits[logits < threshold] = -inf   (mask unlikely tokens)
5. Let existing vLLM sampler handle temperature, softmax, and sampling
```

The standard deviation directly measures the "peakiness" of the logit distribution:
- **Sharp distribution** (model is certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model is uncertain): large std → wide threshold → more candidates

### How it compares to other truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative probability ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |
| temperature | Logit | logits / T — rescales distribution steepness | ❌ No truncation |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — top_nσ filters in logit space first, then softmax only processes the surviving candidates (~2.5% of vocab). In contrast, top_p/min_p must compute a full-vocab softmax first, then truncate in probability space.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula: sharp distributions auto-narrow, flat ones auto-widen. top_k fixes the count; top_p adapts the count but not the threshold tightness.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally), while top_p/min_p candidate counts vary significantly with T.
- **Argmax-invariant** — the top-1 token always survives (0 std from itself), so greedy decoding can skip this processor entirely.

### Usage (via custom logits processor)

```python
from vllm import SamplingParams

sampling_params = SamplingParams(
    temperature=0.8,
    extra_args={"top_n_sigma": 2.0}
)
```

The custom logit processor reads `top_n_sigma` from `extra_args` and applies truncation before the standard sampling pipeline.

### Safety / Edge Cases

| Case | Behavior |
|------|----------|
| `n <= 0` | Reject (raise `ValueError`) |
| `std == 0` (all logits equal) | Keep all tokens (no filtering) |
| NaN/Inf logits | Preserve existing behavior (no-op) |
| After filtering | **At least the argmax token is always kept** (prevent all-`-inf`) |

### Comparison Example

```python
# Logit distribution: max=10.0, std=2.0

top_k=5:           Keeps exactly 5 tokens
top_p=0.9:         Keeps tokens summing to 90% cumulative probability
min_p=0.05:        Keeps tokens with prob >= 5% of max
top_n_sigma=2.0:   Keeps tokens with logit >= 10.0 - 2*2.0 = 6.0
```

### Deliverables (if maintainers are interested)

1. **Example file**: `examples/features/logits_processor/top_n_sigma.py`
2. **Unit tests**: Verify masking, edge cases, and compatibility with other samplers
3. **Documentation**: One-paragraph note in sampling docs

---

**References**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### PR DESCRIPTION
**Pull request by Jackie2049**
## Summary

Closes #45023

Adds an example demonstrating **Top-nσ** as a custom logits processor in vLLM, based on the paper:

> Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### What is Top-nσ?

Top-nσ is a logit-space dynamic truncation method that uses the standard deviation of the logit distribution to set a filtering threshold:

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike probability-space filters (top_p, min_p), it operates **before softmax** and adapts to distribution "peakiness":
- **Sharp distribution** (model certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model uncertain): large std → wide threshold → more candidates

## Changes

- New file: `examples/features/logits_processor/top_n_sigma.py`
- `apply()` method uses vectorized Torch operations (not Python per-row loop)

### Key properties
- **Argmax-invariant**: doesn't affect greedy decoding
- Uses vLLM's `extra_args` mechanism (no `SamplingParams` changes needed)
- Handles edge cases: NaN/Inf logits, zero std, n≤0 validation

### Usage

```python
from vllm import LLM, SamplingParams

llm = LLM(
    model="facebook/opt-125m",
    logits_processors=[TopNSigmaLogitsProcessor],
)

sampling_params = SamplingParams(
    temperature=0.8,
    extra_args={"top_n_sigma": 2.0}
)
```

## Experiment Report

Full algorithm effectiveness and vectorized implementation verification (7 experiments, RTX 4090):

👉 [**View rendered HTML report**](https://htmlpreview.github.io/?https://raw.githubusercontent.com/Jackie2049/rollout-infra/main/notebook/experiments/top-n-sigma-effectiveness-report.html)

## Testing plan

- [x] Example runs with `facebook/opt-125m`
- [x] Compare output quality vs `top_p`/`min_p` on diverse prompts
  - Tested across sharp/medium/flat logit distributions (V=32000)
  - Top-nσ(n=2.0) produces comparable filtering rates to top_k(50) and min_p(0.05) at typical LLM scenarios (~2.5% of vocab surviving)
  - Temperature stability: Top-nσ maintains consistent candidate count across T=0.1–10.0 (358 tokens at all T), while top_p/min_p vary significantly with temperature scaling
  - Argmax-invariant verified across 32 configurations (all dtype/sharpness/n combinations)
- [x] Benchmark: throughput impact of the per-row Python loop
  - Original row-by-row loop replaced with vectorized Torch ops
  - RTX 4090 benchmark (V=32000): **10.3x** @ B=32, **45.4x** @ B=128, **65.7x** @ B=256
  - Vectorized is bit-for-bit identical to row-by-row across all configs (FP32/FP16)
  - Edge cases (NaN/Inf, std=0, empty req_info) handled identically

### CODE REVIEW

**Comment by Jackie2049** [comment]
Recreating with the correct branch from the fork
